replace Env service with nestjs ConfigService

fmagno opened the pull request:
WIP: I was trying to use simple-auth and noticed that an environment variable was missing in the README file example.
I also noticed the current environment service could be made a little bit more declarative and take advantage of the validation mechanism of the nestjs/ConfigModule.

It still requires testing! :)

fmagno committed:
replace Env service with nestjs ConfigService

fmagno committed:
Fix cast while getting DB_SESSIONS env value

GJordao commented:
I can't seem to be able to run this locally, its giving an error on all the non string env variables. I'm assuming that it probably doesn't convert them from string?

I didn't know about Nest's config service, it looks great but does it give you auto complete for the variables? The idea of the service was mostly because of that 😅

GJordao commented:
screenshot for reference

![image](https://user-images.githubusercontent.com/21367039/117059565-d8ea0300-ad17-11eb-82d9-32821820b6a7.png)

fmagno committed:
Fix/add env variables validation

fmagno commented:
Gotcha! That issue should be fixed, now.

If you think env variables auto-complete is really a must then it is possible to create a thin wrapper service around `ConfigService` with a getter function for each variable - take a look [here](https://docs.nestjs.com/techniques/configuration#custom-getter-functions).

Issue #15 should be covered as well.

fmagno committed:
wip: add unit/e2e tests and allow setting env vars programatically

fmagno commented:
WIP:

Converted the AppModule to a Dynamic module to allow passing environment variables programmatically.
Implemented a few unit tests to test the validation function.
Also added a few e2e tests to test the registration/login process. I will be adding some more soon.

fmagno committed:
Add missing devDependency sqlite

fmagno committed:
Update package-lock

fmagno committed:
Fix unit test (DATABASE_TYPE=sqlite)

fmagno committed:
EnvValidation refactoring

MicroAnibal commented:
Welcome @fmagno, 
Thanks for contributing :+1:

> Gotcha! That issue should be fixed, now.

Confirmed, is fixed, solution is working locally.

> If you think env variables auto-complete is really a must then it is possible to create a thin wrapper service around `ConfigService` with a getter function for each variable - take a look [here](https://docs.nestjs.com/techniques/configuration#custom-getter-functions).

Agree :100: with @GJordao, auto-complete is a must have to speed things up and save some brain cycles  
Also agree that the thin wrapper is perfect for this case.

fmagno committed:
Add a thin wrapper (ConfigServiceApi) around ConfigService to allow autocomplete

fmagno commented:
Hey @MicroAnibal! Cheers! :)
Autocomplete should now work with the service ConfigServiceApi.

GJordao committed:
Merge branch 'master' into wip/replace-env-service-with-nestjs-config

GJordao commented on src/AppModule.ts:
```diff
@@ -1,60 +1,87 @@
+                    useFactory: (configApi: ConfigServiceApi) =>
+                        ({
+                            type:
+                                configApi.DATABASE_TYPE === 'test'
```
Why are we validating for a possible "test" database type here?

GJordao commented on src/Config/decorators.ts:
```diff
@@ -0,0 +1,88 @@
+                return false;
+            }
+            if (
+                databaseType === 'test' &&
```
Same question here

GJordao commented on src/Config/decorators.ts:
```diff
@@ -0,0 +1,88 @@
+    ValidationOptions
+} from 'class-validator';
+
+export function IsDefinedAndNotEmptyIfIsMysqlOrPostgres(
```
We'll support more DB types in the near future (typeORM supports a couple by default we just haven't tested it 😅 ). Maybe we should rename the function to `IsDefinedAndNotEmptyIfValidDbType`?

GJordao commented on src/testing/RegistrationTestApp.ts:
```diff
@@ -0,0 +1,24 @@
+import * as request from 'supertest';
```
Can we move all the tests to `/src/__tests__` just to keep everything grouped an in the same place?

GJordao commented on test/registration.e2e-spec.ts:
```diff
@@ -0,0 +1,96 @@
+import { JWT_PATTERN } from '../src/testing';
```
Maybe we could move this folder to `/src/__tests___/e2e`

GJordao commented on src/Config/test/EnvValidation.spec.ts:
```diff
@@ -0,0 +1,65 @@
+import { validate } from '../validation';
```
Can we move the test to `/src/__tests__`?

GJordao commented on src/Config/validation.ts:
```diff
@@ -0,0 +1,125 @@
+import {
+    IsDefinedAndNotEmptyIfIsMysqlOrPostgres,
+    IsPortIfIsMysqlOrPostgres
+} from './Decorators';
```
Should be `from './decorators'` with lowercase `d`

GJordao commented on src/Config/ConfigModule.ts:
```diff
@@ -0,0 +1,10 @@
+import { Module } from '@nestjs/common';
```
Could we rename this to EnvModule perhaps? I find the naming confusing. Also could we move it to inside `/src/configs`. Trying to avoid having a lot of clutter in the `/src/` folder. I think long term it will make the app easier to navigate. Thoughts?
